### PR TITLE
H-1015: Add TypeScript entity inference Temporal workflow, REST endpoint

### PR DESCRIPTION
**Pull request by CiaranMn**
## 🌟 What is the purpose of this PR?

Writes a TypeScript Temporal workflow to infer entities of requested types from provided text, and exposes it via a REST endpoint.

The motivation for doing this is:
1. It will be easier to maintain and expand the worker in TypeScript rather than Python as it is currently, and easier to share types
2. Exposing it via a REST endpoint means we can keep the connection alive to avoid request timeouts (Cloudflare has a 100 second timeout, and the GraphQL query isn't designed to send chunks to keep the connection alive)

Functionality differences from the existing Python workflow:
1. It handles inheritance
2. It creates a combined schema for requested types to pass to the AI model, thus providing property type descriptions and handling non-primitive property type values (e.g. property type objects)
3. It makes one request to the AI model for entities of all types to be inferred, rather than sequential requests per type (seems to work and is faster, but may need to tweak / revisit based on further testing – not a big deal to change back to sequential requests per type)
4. Uses the new 'validateEntity' functionality in the Graph API to validate entities before attempting to create them
5. Creates entities rather than returning proposed entities for the user to select from – in future, these will be created as 'draft entities' which the user can discard as desired.
6. Removes `allowEmptyResults` argument – happy to put this back but can't remember why we had it, the only functional difference it makes is the caller specifying whether empty results should cause an error or not. Not sure that's necessary – we can decide whether empty results counts as an error and throw it if we want to be able to track them
7. updates the base TS config's lib to `es2022` from `es2020` (required to be able to specify additional arguments to the `Error` constructor, to pass the `Status` objects as `error.cause`)

## 🚫 Blocked by

- [ ] Testing

## 🔍 What does this change?

- Adds an `inferEntities` workflow to the existing TS worker, and removes the unused workflows present in it
- expose types for the workflow's arguments/return in a shared package (means the API has access to the types without having to import the workflow's function, and means the browser extension can use the types at all)
- moves `createGraphClient` to a shared package to avoid duplicating code where one is needed
- removes `@apps/hash-api` as a dependency of the TS worker

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## 🐾 Next steps

1. Update browser extension to use the new REST endpoint, and to remove the 'select proposed entities' step – instead, just show which entities were created, grouped by page inferred from
2. Remove the Python Temporal worker, and the `inferEntities` GraphQL mutation

## 🛡 What tests cover this?

- I added a test for the 'combine schema' function, otherwise manual

## ❓ How to test this?

1. Set `OPENAI_API_KEY` in your `env.local`
1. Run the Node API (`yarn dev:backend`)
1. Get your `ory_kratos_session` cookie from your browser for your local app (you'll need to run `yarn dev:frontend` and log in if you don't have an authenticated session cookie)
1. Create some entity type(s) to test with
1. Get the `ownedById` of the web you want to create entities in
1. Stop the `hash-ai-worker-ts` in Docker and run `yarn workspace @apps/hash-ai-worker-ts` (or rebuild `external-services`, but that takes much longer)
1. Make a `POST` request to `http://localhost:5001/entities/infer` with the required body (see demo below)
1. Try with some garbage text or other bad params to check error handling

## 📹 Demo

<details>
<summary>Input</summary>

```json
{
    "entityTypeIds": [
        "http://localhost:3000/@alice/types/entity-type/person/v/2",
        "http://localhost:3000/@alice/types/entity-type/company/v/1",
        "http://localhost:3000/@alice/types/entity-type/employed-by/v/2"
    ],
    "temperature": 0,
    "maxTokens": 4096,
    "model": "gpt-4-1106-preview",
    "ownedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d",
    "validation": "FULL",
    "textInput": "Microsoft Corporation is an American multinational technology corporation headquartered in Redmond, Washington. Microsoft's best-known software products are the Windows line of operating systems, the Microsoft 365 suite of productivity applications, and the Edge web browser. Its flagship hardware products are the Xbox video game consoles and the Microsoft Surface lineup of touchscreen personal computers. Microsoft ranked No. 14 in the 2022 Fortune 500 rankings of the largest United States corporations by total revenue;[2] it was the world's largest software maker by revenue as of 2022. It is considered one of the Big Five American information technology companies, alongside Alphabet (parent company of Google), Amazon, Apple, and Meta Platforms.Microsoft was founded by Bill Gates and Paul Allen on April 4, 1975, to develop and sell BASIC interpreters for the Altair 8800. It rose to dominate the personal computer operating system market with MS-DOS in the mid-1980s, followed by Windows. The company's 1986 initial public offering (IPO) and subsequent rise in its share price created three billionaires and an estimated 12,000 millionaires among Microsoft employees. Since the 1990s, it has increasingly diversified from the operating system market and has made several corporate acquisitions, the largest being the acquisition of LinkedIn for $26.2 billion in December 2016,[3] followed by its acquisition of Skype Technologies for $8.5 billion in May 2011.[4]. As of 2015, Microsoft is market-dominant in the IBM PC compatible operating system market and the office software suite market, although it has lost the majority of the overall operating system market to Android.[5] The company also produces a wide range of other consumer and enterprise software for desktops, laptops, tabs, gadgets, and servers, including Internet search (with Bing), the digital services market (through MSN), mixed reality (HoloLens), cloud computing (Azure), and software development (Visual Studio).Steve Ballmer replaced Gates as CEO in 2000 and later envisioned a 'devices and services' strategy.[6] This unfolded with Microsoft acquiring Danger Inc. in 2008,[7] entering the personal computer production market for the first time in June 2012 with the launch of the Microsoft Surface line of tablet computers, and later forming Microsoft Mobile through the acquisition of Nokia's devices and services division. Since Satya Nadella took over as CEO in 2014, the company has scaled back on hardware and instead focused on cloud computing, a move that helped the company's shares reach their highest value since December 1999.[8][9]. Earlier dethroned by Apple in 2010, in 2018, Microsoft reclaimed its position as the most valuable publicly traded company in the world.[10] In April 2019, Microsoft reached a trillion-dollar market cap, becoming the third U.S. public company to be valued at over $1 trillion after Apple and Amazon, respectively. As of 2022, Microsoft has the fourth-highest global brand valuation.Microsoft has been criticized for its monopolistic practices and the company's software has been criticized for problems with ease of use, robustness, and security."
}
```
</details>

<details>
<summary>Output</summary>

```json
{
    "code": "OK",
    "contents": [
        {
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~8b8c1853-8750-4b1d-a69a-4415c88fdaff",
                    "editionId": "d7730517-dd8c-479c-b05e-7708c7b8c532"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.791223000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.791223000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/company/v/1",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {
                "https://blockprotocol.org/@blockprotocol/types/property-type/display-name/": "Microsoft Corporation",
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "An American multinational technology corporation headquartered in Redmond, Washington, known for the Windows operating systems, Microsoft 365 suite, Edge web browser, Xbox video game consoles, and Microsoft Surface computers. Ranked No. 14 in the 2022 Fortune 500 rankings and the world's largest software maker by revenue as of 2022. Founded by Bill Gates and Paul Allen on April 4, 1975."
            }
        },
        {
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~db82e80c-d869-431e-ace8-517dfd37ca6d",
                    "editionId": "106e9836-0ccf-461b-a9bd-ef5810851458"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.783026000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.783026000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/person/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {
                "https://hash.ai/@hash/types/property-type/preferred-name/": "Bill Gates",
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "Co-founder of Microsoft Corporation, involved in the development and sale of BASIC interpreters for the Altair 8800, and played a key role in the dominance of the personal computer operating system market with MS-DOS and Windows."
            }
        },
        {
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~a5bb6a40-dbbe-4089-9ae5-27f3deafeca2",
                    "editionId": "2fb00333-c232-4ece-834e-fa46c3220676"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.781423000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.781423000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/person/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {
                "https://hash.ai/@hash/types/property-type/preferred-name/": "Paul Allen",
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "Co-founder of Microsoft Corporation, involved in the development and sale of BASIC interpreters for the Altair 8800, and contributed to the company's early success in the personal computer operating system market."
            }
        },
        {
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~0c5b95a8-b544-4e68-a982-4042aab8b037",
                    "editionId": "79bef992-fe08-4c9d-adb5-5e69c15d562a"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.804286000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.804286000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/person/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {
                "https://hash.ai/@hash/types/property-type/preferred-name/": "Steve Ballmer",
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "Succeeded Bill Gates as CEO of Microsoft Corporation in 2000, led the company with a 'devices and services' strategy, and oversaw several acquisitions including Danger Inc. and the entry into the personal computer production market with the Surface tablets."
            }
        },
        {
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~47965b21-0e4c-4c50-ae87-5da2920b661c",
                    "editionId": "5742ee07-90ef-4d62-af8a-e7dc6189e434"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.784626000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.784626000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/person/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {
                "https://hash.ai/@hash/types/property-type/preferred-name/": "Satya Nadella",
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "Became CEO of Microsoft Corporation in 2014, shifted the company's focus towards cloud computing, which contributed to the company's shares reaching their highest value since December 1999."
            }
        },
        {
            "linkData": {
                "leftEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~db82e80c-d869-431e-ace8-517dfd37ca6d",
                "rightEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~8b8c1853-8750-4b1d-a69a-4415c88fdaff"
            },
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~2b93c112-b171-42ed-96d0-34134c36e3ac",
                    "editionId": "394c2b21-692d-41d5-b149-9dd0c54b354e"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.885755000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.885755000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/employed-by/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {}
        },
        {
            "linkData": {
                "leftEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~a5bb6a40-dbbe-4089-9ae5-27f3deafeca2",
                "rightEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~8b8c1853-8750-4b1d-a69a-4415c88fdaff"
            },
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~d81be7fa-3872-4272-bd91-df89d05d98ee",
                    "editionId": "9d5d1346-0c3d-4a9b-9a40-b3ed43a60549"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.882124000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.882124000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/employed-by/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {}
        },
        {
            "linkData": {
                "leftEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~0c5b95a8-b544-4e68-a982-4042aab8b037",
                "rightEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~8b8c1853-8750-4b1d-a69a-4415c88fdaff"
            },
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~6cf1e1ce-2bec-486c-8b5e-06ceb9a409d8",
                    "editionId": "82f03f2b-4cce-4416-8452-a2aaa6b20e5d"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.885532000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.885532000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/employed-by/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {}
        },
        {
            "linkData": {
                "leftEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~47965b21-0e4c-4c50-ae87-5da2920b661c",
                "rightEntityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~8b8c1853-8750-4b1d-a69a-4415c88fdaff"
            },
            "metadata": {
                "recordId": {
                    "entityId": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d~4ce32c1d-3acb-4a3c-8db1-e4ed54063797",
                    "editionId": "31a1c6fb-79ab-4b30-932b-39e1a949dfab"
                },
                "temporalVersioning": {
                    "decisionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.882680000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    },
                    "transactionTime": {
                        "start": {
                            "kind": "inclusive",
                            "limit": "2023-11-27T14:50:20.882680000Z"
                        },
                        "end": {
                            "kind": "unbounded"
                        }
                    }
                },
                "entityTypeId": "http://localhost:3000/@alice/types/entity-type/employed-by/v/2",
                "provenance": {
                    "recordCreatedById": "46627d9c-a4e8-4050-88c0-1b321b8f2e1d"
                },
                "archived": false
            },
            "properties": {}
        }
    ],
    "message": "Created 9 entities, with 0 failures."
}
```
</details>

### CODE REVIEW

File: apps/hash-frontend/src/pages/shared/entity-type-page.tsx [hunk 1]
```diff
@@ -122,7 +122,7 @@ export const EntityTypePage = ({
   const isDraft = !!draftEntityType;
 
   const handleSubmit = wrapHandleSubmit(async (data) => {
-    if (!isDirty && isDraft) {
+    if (!isDirty && !isDraft) {
```
Comment (CiaranMn) [inline apps/hash-frontend/src/pages/shared/entity-type-page.tsx:125]: Driveby fix (this is what was preventing creation of draft types with no properties)

File: apps/hash-ai-worker-ts/src/activities/infer-entities/create-entities.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  requestedEntityTypes: Record<
+    VersionedUrl,
+    { isLink: boolean; schema: DereferencedEntityType }
+  >;
+  ownedById: OwnedById;
+}): Promise<{
+  createdEntities: Entity[];
+  creationFailures: InferEntitiesCreationFailure[];
+}> => {
+  const nonLinkTypes = Object.values(requestedEntityTypes).filter(
```
Comment (benwerner01) [inline apps/hash-ai-worker-ts/src/activities/infer-entities/create-entities.ts:33]: Might be more clear that we're not referring to property or data types here:

```suggestion
  const nonLinkEntityTypes = Object.values(requestedEntityTypes).filter(
```

**Review by benwerner01 — APPROVED** [review]
Great work, thank you Ciaran - have tested locally and only spotted one non-blocking nit whilst going through the code 👍

File: apps/hash-ai-worker-ts/src/activities/infer-entities/create-entities.ts (hunk 2 above)
Comment (CiaranMn) [reply apps/hash-ai-worker-ts/src/activities/infer-entities/create-entities.ts]: Sure – done in https://github.com/hashintel/hash/pull/3630/commits/de29e970691434cf5bd81c447811e1a14a1c8853

**Review by benwerner01 — APPROVED** [review]
